### PR TITLE
Added Opticks hits code

### PR DESCRIPTION
**Pull request by ggalgoczi**

### CODE REVIEW

File: geom/opticks_raindrop.gdml [hunk 1]
```diff
@@ -80,8 +93,8 @@
       <fraction n="0.996003" ref="Ar400x560299384330"/>
     </element>
     <material name="G4_AIR0x560299383bc0" state="gas">
-      <property name="RINDEX" ref="RINDEX0x560299349730"/>
-      <property name="GROUPVEL" ref="GROUPVEL0x560299349110"/>
+      <property name="RINDEX" ref="RINDEX0x5602993a1df1"/>
+      <property name="GROUPVEL" ref="GROUPVEL0x56029938bb90"/>
```
Comment (plexoos) [inline geom/opticks_raindrop.gdml:97]: What was the reason for changing this properties for G4_AIR? Was it needed to see hits?

Comment (plexoos) [reply geom/opticks_raindrop.gdml:97]: So, just to confirm this change is not relevant for getting G4 hits, correct?

File: geom/opticks_raindrop.gdml [hunk 2]
```diff
@@ -109,16 +122,16 @@
...
       <T unit="K" value="293.15"/>
       <MEE unit="eV" value="823"/>
       <D unit="g/cm3" value="11.35"/>
       <fraction n="1" ref="Pb0x560299349da0"/>
     </material>
     <material name="VACUUM0x5602990cb360" state="gas">
-      <property name="RINDEX" ref="RINDEX0x5602993a1df0"/>
-
+      <property name="RINDEX" ref="RINDEX0x5602993a1df3"/>
+<property name="GROUPVEL" ref="GROUPVEL0x56029938bb90"/>
```
Comment (plexoos) [inline geom/opticks_raindrop.gdml:134]: Are these new lines needed?

File: geom/opticks_raindrop.gdml [hunk 3]
```diff
@@ -109,16 +122,16 @@
       <fraction n="0.524" ref="Pb2080x560299383420"/>
     </element>
     <material name="G4_Pb0x5602993831c0" state="solid">
-          <property name="RINDEX" ref="RINDEX0x560299349070"/>
-
+          <property name="RINDEX" ref="RINDEX0x5602993a1df2"/>
+<property name="GROUPVEL" ref="GROUPVEL0x56029938bb90"/>
```
Comment (plexoos) [inline geom/opticks_raindrop.gdml:126]: Same question. Do we need to define a GROUPVEL property to see hits in this volume?

File: geom/opticks_raindrop.gdml [hunk 4]
```diff
@@ -1,12 +1,23 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no" ?>
-<gdml xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="">
+<gdml_simple_extension xmlns:gdml_simple_extension="http://www.example.org"
+    xmlns:xs="http://www.w3.org/2001/XMLSchema-instance"
+    xs:noNamespaceSchemaLocation="SimpleExtension.xsd">
```
Comment (plexoos) [inline geom/opticks_raindrop.gdml:4]: What is the point of switching to this new tag, gdml_simple_extension?

File: geom/opticks_raindrop.gdml (hunk 2 above)
Comment (ggalgoczi) [reply geom/opticks_raindrop.gdml:134]: Rindex is needed for all materials. I have tried running without refractive index before and there were no hits. I am not sure if GROUPVEL is needed or not, some GDML uses it, some not.

File: geom/opticks_raindrop.gdml (hunk 1 above)
Comment (ggalgoczi) [reply geom/opticks_raindrop.gdml:97]: Correct, for G4Hits irrelevant. But we want Optickshits too.

File: geom/opticks_raindrop.gdml (hunk 3 above)
Comment (ggalgoczi) [reply geom/opticks_raindrop.gdml:126]: I think they are not a must.

File: geom/opticks_raindrop.gdml (hunk 4 above)
Comment (ggalgoczi) [reply geom/opticks_raindrop.gdml:4]: Worked in CaTS, copied here. Not sure if has effect.

**Comment by plexoos** [comment]
@ggalgoczi I tried to run the code like this:

```
GEOM=opticks_raindrop ./build-esi-g4ox/src/simg4ox -g esi-g4ox/geom/opticks_raindrop.gdml -m esi-g4ox/run.mac
```

but I don't see anything related to G4 hits in the output. Am I missing something?

**Comment by plexoos** [comment]
Hm... the problem is that I am getting 0 hits, in the output:

```
938                                                                                                                          
939 ==================================================================                                                       
940                                                                                                                          
941 ### Run 0 starts.                                                                                                        
942 PhotonSD::Initialize:  G4_WATER_lv_PhotonDetector   G4_WATER_lv_PhotonDetector_HC                                        
943 PhotonSD::EndOfEvent Number of PhotonHits: 0                                                                             
944 Opticks: NumCollected:  0                                                                                                
945 Opticks: NumCollected:  0                                                                                                
946 Opticks: NumHits:  0                                                                                                     
947  Run terminated.                                                                                                         
948 Run Summary                                                                                                              
949   Number of events processed : 1    
```

**Comment by plexoos** [comment]
I am using the main branch

**Comment by ggalgoczi** [comment]
Please change the photon gun position to 0,0,0 in simtox. If you do not get hits with that let me know and I will take a look!

**Comment by plexoos** [comment]
Yes! Now I see 100 hits in the log. So, here are the two cases:

Default torch, no hits:

Screenshot 2024-11-26 at 4 46 50 PM

0,0,0 torch, 100 hits:

Screenshot 2024-11-26 at 4 51 42 PM

**Comment by plexoos** [comment]
Looks like the only relevant property for getting G4 hits is this: `<auxiliary auxtype="SensDet" auxvalue="PhotonDetector" />`, which we add to a volume.

**Comment by plexoos** [comment]
...and the light has to be inside the volume to be registered on the inner surface as hits. With this configuration I register 29 hits in Pb volume:

Screenshot 2024-11-26 at 7 37 44 PM

**Comment by ggalgoczi** [comment]
Great! So did I assign the sensitiveness to the wrong surface? Or the light did not penetrate the surface?

**Comment by plexoos** [comment]
I am just playing with material properties to get a better feeling. A couple more findings:

- When we define a volume with those auxtype/value, the light does not exit the volume
- Interestingly, it looks like we don't need to define optical surface/border to get reflections like this:

Screenshot 2024-11-26 at 8 41 15 PM

**Comment by ggalgoczi** [comment]
> Looks like the only relevant property for getting G4 hits is this: `<auxiliary auxtype="SensDet" auxvalue="PhotonDetector" />`, which we add to a volume.

It is me who is using this info to identify the sensitive volume in the GDMLParsing step:
```

const G4GDMLAuxMapType *auxmap = parser_.GetAuxMap();
        for (auto const &[logVol, listType] : *auxmap)
        {
            for (auto const &auxtype : listType)
            {
                if (auxtype.type == "SensDet")
                {
                    G4cout << "Attaching sensitive detector to logical volume: " << logVol->GetName() << G4endl;
                    G4String name = logVol->GetName() + "_PhotonDetector";
                    PhotonSD *aPhotonSD = new PhotonSD(name);
                    SDman->AddNewDetector(aPhotonSD);
                    logVol->SetSensitiveDetector(aPhotonSD);
                }
            }
        }
```

Similar code is built into Opticks.

**Comment by ggalgoczi** [comment]
> ...and the light has to be inside the volume to be registered on the inner surface as hits. With this configuration I register 29 hits in Pb volume:
> 
> Screenshot 2024-11-26 at 7 37 44 PM

I suspect these are G4Hits and not OpticksHits, correct?

**Comment by plexoos** [comment]
> I suspect these are G4Hits and not OpticksHits, correct?

Yes, these are Geant4 hits. I still need to learn from you how to extract Opticks hits.

**Comment by ggalgoczi** [comment]
> > I suspect these are G4Hits and not OpticksHits, correct?
> 
> Yes, these are Geant4 hits. I still need to learn from you how to extract Opticks hits.

When I populate the GPU GenSteps I use `U4::CollectGenstep_G4Cerenkov_modified` and then I do get OpticksHits.

**Comment by ggalgoczi** [comment]
So I suspect it is about how we pass the photons to the GPU. Now I tried to increase the number of photons on the GPU to 100k and the runtime of the GPU code did not increase. Either the overhead is so small or we do not pass the photons to the GPU.

**Comment by ggalgoczi** [comment]
Found the solution. Need to pass the created photons to Opticks. Until now we only simulated them with G4.

`U4VPrimaryGenerator::GeneratePrimaries_From_Photons`

**Comment by ggalgoczi** [comment]
Okay, I was wrong! Just needed to change 

`gx->simulate(eventID, true);`

to

`gx->simulate(eventID, false);`

No need for `U4VPrimaryGenerator`

**Comment by ggalgoczi** [comment]
However in G4CXApp.h the following code is used to simulate optical photons. @plexoos  can you comment where you got the code that you use in the main branch?

```
NP* ph = sev->getInputPhoton(); 
U4VPrimaryGenerator::GeneratePrimaries_From_Photons(event, ph) ;
```

File: src/g4app.h [hunk 5]
```diff
@@ -302,7 +336,30 @@ struct EventAction : G4UserEventAction
...
+        std::cout << "Opticks: NumCollected:  " << SEvt::GetNumPhotonCollected(0) << std::endl;
+
+        std::cout << "Opticks: NumHits:  " << num_hits << std::endl;
+        if (num_hits > 0)
+        {
+            G4HCtable *hctable = G4SDManager::GetSDMpointer()->GetHCtable();
+            for (G4int i = 0; i < hctable->entries(); ++i)
+            {
+                std::string sdn = hctable->GetSDname(i);
+                std::size_t found = sdn.find("Photondetector");
```
[the remark this answers is not in the document]
Comment (ggalgoczi) [reply src/g4app.h:354]: Ohh yes, that part was not tested yet, that is why I did not merge it in yet. It is good that you found it, thank you!

**Pull request merged**